lib/api/core/sdk/funnelProtocol.js

```diff
@@ -22,23 +22,34 @@
 'use strict';
 
 const
-  KuzzleEventEmitter = require('kuzzle-sdk').KuzzleEventEmitter,
-  Request = require('kuzzle-common-objects').Request;
+  Request = require('kuzzle-common-objects').Request,
+  KuzzleEventEmitter = require('kuzzle-sdk').KuzzleEventEmitter;
 
 class FunnelProtocol extends KuzzleEventEmitter {
-  constructor(funnel) {
+  constructor(funnel, originalRequest = null) {
```

Comment:
Instead of `originalRequest` I would use a `RequestContext` object as second parameter:

If we want to inject a custom context inside a Hook or a Pipe, we don't want to instantiate a full dummy Request as all we want is the context.

So the SDK instantiation would be:
* without context:
```js
this.context.accessors.sdk.document.create(...)
```

* inside a plugin controller, if we want the context of the original request:
```js
this.context.accessors.sdk.context(request.context).document.create(...)
```

* inside a Hook or a Pipe, if we want to inject a custom context (for example: user = my-plugin-identifier):
```js
const customContext = new RequestContext({my-custom-context-info});
this.context.accessors.sdk.context(customContext).document.create(...);
```

Reply:
That means we should also expose a RequestContext contructor to PluginContext...

Reply:
For me this syntax is a little heavy: `this.context.accessors.sdk.context(request.context).document.create`  

I'm thinking about two solutions:
  
  - `context()` can take either a Request or a RequestContext
  - `context()` always take a Request, even if we didn't need a full dummy Request, it's much simpler to understand since developers are more used to work with Request objects than RequestContext objects directly.

Reply:
> For me this syntax is a little heavy: `this.context.accessors.sdk.context(request.context).document.create`

Really more heavy than `this.context.accessors.sdk.context(request).document.create` ?

>     * `context()` can take either a Request or a RequestContext

=> NO WAY It's always a bad solution to allow different types for a single parameter.
 
>     * `context()` always take a Request, even if we didn't need a full dummy Request, it's much simpler to understand since developers are more used to work with Request objects than RequestContext objects directly.

But:
* the Request object contains lots of attributes that are totally unused here 
* developers are not much used to **create** Request objects (and which `controller`, `action`, etc, should we give to the constructor here ???)
* it's not less easy to understand how to create a RequestContext object with a `user` attribute

Reply:
A little bit, but I admit it doesn't really change much...

After some thought I agree that passing a RequestContext object is more logic.  

The point is that for most of the request we might want to use the original user id and for some we want to use a custom user id.
Maybe we can pass directly an object representing the user ?

```
this.context.accessors.sdk.context(request.context.user).document.create(...)

const user = { _id: 'admin' }
this.context.accessors.sdk.context(user).document.create(...)
```

In this case maybe we can think about another method name than `context`

Reply:
about method name, see also my other suggestion below, about exposing a sdk constructor instead of an accessor.

about the type of the "contextual" parameter, what do we really need (what from the request's context is relevant ?)
Actually, the RequestContext contains following attributes:
* `token`: the current's user JWT => not usable in a plugin context
* `connection`: object representing the original network connection (as well as the protocol name) => no need to transmit the original request's `connection` object, but we could inject automatically in each request a custom object with "internal" as protocol name, for sample ?
* `user`: object representing the current logged in user... do we need the whole object, or only his ID ? What if we inject a custom user ID that do not exist inside our database ? Do we accept it anyway or do we need to validate his existence ? (I really do not know what is the best approach, but we should ask ourselves this question).

Reply:
... To resume, I suggest, as you, that we only need the `user` from the context (the whole object or his ID... to be decided), and we could also add a custom `connection.protocol` object to the "internal" RequestContext, like this:
```js
class FunnelProtocol extends KuzzleEventEmitter {
  constructor(funnel, user = null) {
    super();

    this.funnel = funnel;
    this.id = 'funnel';
    this.user = user;
  }

  isReady () {
    return true;
  }

  query (request) {
    const kuzzleRequest = new Request(request, {protocol: "funnel"});

    if (this.user) {
      kuzzleRequest.context.user = this.user;
    }

    return this.funnel.executePluginRequest(kuzzleRequest)
      .then(result => ({ result }));
  }
}
```

Reply:
I agree with this last proposal, passing the complete user. Only, I would then rename `context` to `as`:

`this.context.accessors.sdk.as(request.context.user).document.create(...)`